irohad/ametsuchi/impl/rocksdb_block_storage.cpp

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ametsuchi/impl/rocksdb_block_storage.hpp"
+
+#include "ametsuchi/impl/rocksdb_common.hpp"
+#include "backend/protobuf/block.hpp"
+#include "common/byteutils.hpp"
+#include "logger/logger.hpp"
+
+using namespace iroha::ametsuchi;
+
+RocksDbBlockStorage::RocksDbBlockStorage(
+    std::shared_ptr<RocksDBContext> db_context,
+    std::shared_ptr<shared_model::interface::BlockJsonConverter> json_converter,
+    logger::LoggerPtr log)
+    : db_context_(std::move(db_context)),
+      json_converter_(std::move(json_converter)),
+      log_(std::move(log)) {}
+
+bool RocksDbBlockStorage::insert(
+    std::shared_ptr<const shared_model::interface::Block> block) {
+  return json_converter_->serialize(*block).match(
+      [&](const auto &block_json) {
+        RocksDbCommon common(db_context_);
+
+        if (auto result =
+                forBlock<kDbOperation::kCheck, kDbEntry::kMustNotExist>(
+                    common, block->height());
+            expected::hasError(result)) {
+          log_->warn(
+              "Error while block {} insertion. Code: {}. Description: {}",
+              block->height(),
+              result.assumeError().code,
+              result.assumeError().description);
+          return false;
+        }
+
+        common.valueBuffer() = block_json.value;
+        if (auto result = forBlock<kDbOperation::kPut>(common, block->height());
+            expected::hasError(result)) {
+          log_->error("Error while block {} storing. Code: {}. Description: {}",
+                      block->height(),
+                      result.assumeError().code,
+                      result.assumeError().description);
+          return false;
+        }
+
+        uint64_t count = 0ull;
+        if (auto result =
+                forBlocksTotalCount<kDbOperation::kGet, kDbEntry::kMustExist>(
+                    common);
+            expected::hasValue(result))
+          count = *result.assumeValue();
+
+        common.encode(count + 1ull);
+        if (auto result =
+                forBlocksTotalCount<kDbOperation::kPut, kDbEntry::kMustExist>(
```

Comment:
Does it make sense to store block itself and block's count atomically?

Reply:
It's atomic. RocksdbCommon ctor make lock.

Reply:
<kDbOperation::kPut, kDbEntry::kMustExist> is not obvious to understand.

Should the record exist before operation or after it. I expect it kMustExist before the operation, so must be rewritten. Please provide more obvious notation.

Reply:
To make check before operation execution is a logic case. I asked you to add issue with extended ```kDbEntry``` checks for ```kPut``` and ```kDel``` operations.

Reply:
#1330